src/appleseed-max-impl/appleseedrenderelement/appleseedrenderelement.cpp

```diff
@@ -0,0 +1,495 @@
+
+//
+// This source file is part of appleseed.
+// Visit http://appleseedhq.net/ for additional information and resources.
+//
+// This software is released under the MIT license.
+//
+// Copyright (c) 2018 Sergo Pogosyan, The appleseedhq Organization
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// Interface header.
+#include "appleseedrenderelement.h"
+
+// appleseed-max headers.
+#include "appleseedrenderelement/resource.h"
+#include "appleseedrenderer/appleseedrenderer.h"
+#include "main.h"
+#include "utilities.h"
+#include "version.h"
+
+// appleseed.renderer headers.
+#include "renderer/api/aov.h"
+
+// 3ds Max headers.
+#include <iparamm2.h>
+
+// Standard headers.
+#include <string>
+#include <vector>
+
+namespace asf = foundation;
+namespace asr = renderer;
+
+namespace
+{
+    const wchar_t* AppleseedRenderElementFriendlyClassName = L"appleseed_AOV";
+
+    enum { ParamBlockIdRenderElement };
+    enum { ParamBlockRefRenderElement };
+
+    enum ParamId
+    {
+        ParamIdEnableOn         = 0,
+        ParamIdElementBitmap    = 1,
+        ParamIdElementName      = 2,
+        ParamIdAOVIndex         = 3,
+    };
+}
+
+AppleseedRenderElementClassDesc g_appleseed_renderelement_classdesc;
+
+//
+// AppleseedRenderElement class implementation.
+//
+
+namespace
+{
+    
+    std::vector<std::string> get_aov_names()
+    {
+        const asr::AOVFactoryRegistrar factory_registrar;
+        auto factories = factory_registrar.get_factories();
+        auto aov_names = std::vector<std::string>();
+        for (size_t i = 0, e = factories.size(); i < e; ++i)
+            aov_names.push_back(factories[i]->get_model_metadata().get("label"));
+
+        return aov_names;
+    }
+
+    void init_combo_box(IParamMap2* param_map)
+    {
+        HWND dlg_hwnd = 0;
```

Comment:
I think that ultimately `HWND` is a pointer so `nullptr` would be more appropriate.